sdk/confidentialledger/azure-confidentialledger/azure/confidentialledger/_shared/async_confidential_ledger_client_base.py

```diff
@@ -0,0 +1,134 @@
+from .credential import ConfidentialLedgerCertificateCredential
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+
+
+class AsyncConfidentialLedgerClientBase(object):
+    def __init__(
+        self,
+        *,
+        endpoint: str,
+        credential: Union[ConfidentialLedgerCertificateCredential, "TokenCredential"],
+        ledger_certificate_path: str,
+        **kwargs: Any
+    ) -> None:
+
+        client = kwargs.get("generated_client")
+        if client:
+            # caller provided a configured client -> nothing left to initialize
+            self._client = client
+            return
+
+        if not endpoint:
+            raise ValueError("Expected endpoint to be a non-empty string")
+
+        if not credential:
+            raise ValueError("Expected credential to not be None")
+
+        if not isinstance(ledger_certificate_path, str):
+            raise TypeError("ledger_certificate_path must be a string")
+
+        if ledger_certificate_path == "":
+            raise ValueError(
+                "If not None, ledger_certificate_path must be a non-empty string"
+            )
+
+        endpoint = endpoint.strip(" /")
+        try:
+            if not endpoint.startswith("https://"):
+                self._endpoint = "https://" + endpoint
+            else:
+                self._endpoint = endpoint
+        except AttributeError as e:
+            raise ValueError("Confidential Ledger URL must be a string.") from e
+
+        self.api_version = kwargs.pop("api_version", DEFAULT_VERSION)
+
+        pipeline = kwargs.pop("pipeline", None)
+        transport = kwargs.pop("transport", None)
+        if transport is None:
+            # Customize the transport layer to use client certificate authentication and validate
+            # a self-signed TLS certificate.
+            if isinstance(credential, ConfidentialLedgerCertificateCredential):
+                # The async version of the client seems to expect a sequence of filenames.
+                # azure/core/pipeline/transport/_aiohttp.py:163
+                # > ssl_ctx.load_cert_chain(*cert)
+                kwargs["connection_cert"] = (credential.certificate_path,)
+
+            kwargs["connection_verify"] = ledger_certificate_path
+            transport = AioHttpTransport(**kwargs)
+
+        http_logging_policy = HttpLoggingPolicy(**kwargs)
+        http_logging_policy.allowed_header_names.update(
+            {
+                "x-ms-keyvault-network-info",
+                "x-ms-keyvault-region",
+                "x-ms-keyvault-service-version",
+            }
+        )
+
+        if not isinstance(credential, ConfidentialLedgerCertificateCredential):
+            kwargs["authentication_policy"] = kwargs.pop(
+                "authentication_policy",
+                AsyncBearerTokenCredentialPolicy(
+                    credential,
+                    "https://confidential-ledger-ppe.azure.com/.default",
```

Comment:
Is hardcoding the `scopes` here recommended style? Or is a `kwargs["scopes"]` the typical pattern?

(Pushing a change to remove the `ppe` part of the url)